### PR TITLE
[Client] Skip client object ref, actor handle, and actor ref dealloc/del if client package has already been cleaned up

### LINKED ISSUE #17647: Empty error at the end of script getting futures from ray server on cluster.
[autoscaler]

### What is the problem?

- python version: 3.7/3.8 (confirmed that it breaks for both)
- `ray` version:  1.5.1
- local OS: MacOS Big Sur 11.4 (local)
- remote OS: ubuntu focal 20.04 (remote)

### Reproduction
I set up a ray cluster which, upon launch, configures itself to be a server (config file to follow). I then run the following script locally.

```
import ray
import ray.util

REMOTE_CLUSTER_IP = '0.0.0.0'
REMOTE_CLUSTER_PORT = 0

ray.util.connect(f'{REMOTE_CLUSTER_IP}:{REMOTE_CLUSTER_PORT}')

@ray.remote
def hello_world():
    return 'Hello, world!'

result = ray.get([hello_world.remote() for _ in range(5)])
if result == ['Hello, world!' for _ in range(5)]:
    print('Got expected result!')

print('Script exiting...')
```

Here is the perplexing stdout I get (verbatim):
```
>>> python script.py
Got expected result!
Script exiting...
Error in sys.excepthook:

Original exception was:

```

The `config.yaml` file I'm using to spin up my cluster is as follows.

```

cluster_name: cluster

max_workers: 2
upscaling_speed: 1.0

docker:
    image: rayproject/ray:nightly-py37 | rayproject/ray:nightly-py38
    container_name: "ray_container"
    pull_before_run: False # pulled down already in AMI (which is rebuilt every day)
    run_options: []

idle_timeout_minutes: 5

# AWS config
provider:
    type: aws
    region: <AWS_REGION>
    cache_stopped_nodes: False

auth:
    ssh_user: ubuntu
    ssh_private_key: <PRIVATE_KEY>

available_node_types:
    ray.head.custom:
        min_workers: 0
        max_workers: 0
        resources: {}
        node_config:
            SubnetId: <SUBNET_ID>
            ImageId: <AMI>
            SecurityGroupIds: <SECURITY_GROUP_IDS>
            InstanceType: <INSTANCE_TYPE>
            BlockDeviceMappings:
                - DeviceName: /dev/sda1
                  Ebs:
                      VolumeSize: 100
    ray.worker.custom:
        min_workers: 0
        max_workers: 2
        resources: {}
        node_config:
            SubnetId: <SUBNET_ID>
            ImageId: <AMI>
            SecurityGroupIds: <SECURITY_GROUP_IDS>
            InstanceType: <INSTANCE_TYPE>
            InstanceMarketOptions:
                MarketType: spot

head_node_type: ray.head.custom

file_mounts: {}
cluster_synced_files: []
file_mounts_sync_continuously: False

rsync_exclude:
    - "**/.git"
    - "**/.git/**"

rsync_filter:
    - ".gitignore"


# forward external dashboard port to internal dashboard port so that dashboard is accessible from external port
initialization_commands: ['sudo iptables -t nat -A PREROUTING -p tcp --dport <DASHBOARD_EXTERNAL_PORT> -j REDIRECT --to-port <DASHBOARD_PORT>']
setup_commands: []
head_setup_commands: []
worker_setup_commands: []

head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --dashboard-host "<DASHBOARD_IP>" --dashboard-port <DASHBOARD_PORT> --head --port=<PORT> --object-manager-port=<OBJECT_MANAGER_PORT> --autoscaling-config=~/ray_bootstrap_config.yaml
    - nohup python -m ray.util.client.server --host "<HOST_IP>" --port <HOST_PORT> &
# Start the worker node.
worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:<PORT> --object-manager-port=<OBJECT_MANAGER_PORT>

head_node: {}
worker_nodes: {}
```

It seem like the shutdown/cleanup process is erroring out in some mysterious way (however, even if I manually shut down before the script exits, I still get the same error, so I have no idea what's actually happening). I think that this is all the information needed to replicate the bug, but please do let me know if there's anything I'm forgetting or if I can elaborate in any way. I also welcome any and all comments about how I'm doing stuff!

### LINKED ISSUE #17968: Confusing error message when ClientObjectRefs are cleaned up on script exit
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray 1.5.2, Python 3.8

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

(Requires Ray client server to be running on port 50051)

```python
import ray

ray.client("localhost:50051").connect()

@ray.remote
def f():
    return 42

@ray.remote
class SomeClass:
    pass

# This object ref will be cleaned up when the script exits
obj_ref = f.remote()
# This actor handle will be cleaned up when the script exits
actor_handle = SomeClass.remote()
```

Results in this error message when the script exits:
```
AttributeError: 'NoneType' object has no attribute 'ray'
Exception ignored in: 'ray._raylet.ClientObjectRef.__dealloc__'
Error in sys.excepthook:

Original exception was:
```
and
```
Exception ignored in: <function ClientActorHandle.__del__ at 0x7fe48f6c7280>
Traceback (most recent call last):
  File "/Users/cwong/ray38/python/ray/util/client/common.py", line 219, in __del__
AttributeError: 'NoneType' object has no attribute 'is_connected'
AttributeError: 'NoneType' object has no attribute 'ray'
Exception ignored in: 'ray._raylet.ClientActorRef.__dealloc__'
AttributeError: 'NoneType' object has no attribute 'ray
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

Check if client package ha been set to None when deallocating client object refs. When `__dealloc__` is called without this check, it errors, leading to `AttributeError` in `ClientObjectRef.__dealloc__`, and sys.excepthook errors when cleaning up ClientRemoteFunc's.

Should fix these confusing messages on script exit:
```
AttributeError: 'NoneType' object has no attribute 'ray'
Exception ignored in: 'ray._raylet.ClientActorRef.__dealloc__'
AttributeError: 'NoneType' object has no attribute 'ray
```

```
Exception ignored in: <function ClientActorHandle.__del__ at 0x7fe48f6c7280>
Traceback (most recent call last):
  File "/Users/cwong/ray38/python/ray/util/client/common.py", line 219, in __del__
AttributeError: 'NoneType' object has no attribute 'is_connected'
```

```
AttributeError: 'NoneType' object has no attribute 'ray'
Exception ignored in: 'ray._raylet.ClientObjectRef.__dealloc__'
```

```
Error in sys.excepthook:

Original exception was:
```

## Related issue number

Closes #17968, Closes #17647
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ijrsvt — APPROVED** [review]
LGTM!

**Comment by ckw017** [comment]
Failures (windows:test_advanced and windows:test_advanced_client) very likely unrelated (both at 20% failure rate, 100% over the last ~10 runs). Reviewers can merge at discretion.

**Review by ijrsvt — CHANGES_REQUESTED** [review]
Sorry for the last-minute change, but can you also modify the [ClientActorRef](https://github.com/ray-project/ray/blob/9416fce91b38d4a028be5ac97e3d0d271a7db378/python/ray/includes/unique_ids.pxi#L316)

File: python/ray/util/client/common.py [hunk 1]
```diff
@@ -216,6 +216,8 @@ def __init__(self,
                                        is_function_or_method)).keys())
 
     def __del__(self) -> None:
+        if ray is None:
```
Comment (AmeerHajAli) [inline python/ray/util/client/common.py:219]: how about client?
Can you please add a comment here please?

File: python/ray/includes/unique_ids.pxi [hunk 2]
```diff
@@ -314,6 +314,8 @@ cdef class ClientActorRef(ActorID):
         client.ray.call_retain(id)
 
     def __dealloc__(self):
+        if client is None or client.ray is None:
```
Comment (AmeerHajAli) [inline python/ray/includes/unique_ids.pxi:317]: Can you please add a comment here?

**Review by AmeerHajAli — APPROVED** [review]

**Comment by ckw017** [comment]
Windows failure unrelated

**Review by ijrsvt — APPROVED** [review]
LGTM! Thanks Chris!

**Comment by ijrsvt** [comment]
Windows failure is unrelated (happening on master as well & on a non-client test): 
Screen Shot 2021-08-20 at 3 18 23 PM

**Pull request merged**